polyfill/lib/absolute.mjs

```diff
@@ -1,126 +1,197 @@
+  }
+
+  plus(durationLike = {}) {
+    const duration = ES.GetIntrinsic('%Temporal.duration%')(durationLike);
+    if (duration.years) throw new RangeError(`invalid duration field years`);
+    if (duration.months) throw new RangeError(`invalid duration field months`);
+
+    let { ms, ns } = GetSlot(this, EPOCHNANOSECONDS);
+    let negative = ms < 0 || ns < 0;
+    ns += duration.microseconds * 1000;
+    ns += duration.nanoseconds;
+    if (negative && ns > 0) {
+      ms += Math.floor(ns / 1e6);
+      ns = 1e6 - (ns % 1e6);
+    }
+    ms += duration.days * 86400000;
+    ms += duration.hours * 3600000;
+    ms += duration.minutes * 60000;
+    ms += duration.seconds * 1000;
+    ms += duration.milliseconds;
 
-Absolute.prototype.getEpochSeconds = function getEpochSeconds() {
-  let epochNanoSeconds = GetSlot(this, EPOCHNANOSECONDS);
-  let epochSecondsBigInt = epochNanoSeconds / 1000000000n;
-  let epochSeconds = ES.ToNumber(epochSecondsBigInt);
-  return epochSeconds;
-};
-Absolute.prototype.getEpochMilliseconds = function getEpochMilliseconds() {
-  let epochNanoSeconds = GetSlot(this, EPOCHNANOSECONDS);
-  let epochMillisecondsBigInt = epochNanoSeconds / 1000000n;
-  let epochMilliseconds = ES.ToNumber(epochMillisecondsBigInt);
-  return epochMilliseconds;
-};
-Absolute.prototype.getEpochMicroseconds = function getEpochMicroseconds() {
-  let epochNanoSeconds = GetSlot(this, EPOCHNANOSECONDS);
-  let epochMicroseconds = epochNanoSeconds / 1000n;
-  return epochMicroseconds;
-};
-Absolute.prototype.getEpochNanoseconds = function getEpochNanoseconds() {
-  let epochNanoSeconds = GetSlot(this, EPOCHNANOSECONDS);
-  return epochNanoSeconds;
-};
+    if (negative && ms > 0) {
+      ms -= 1;
+      ns += 1e6;
+    }
 
-Absolute.prototype.plus = function plus(durationLike = {}) {
-  const Duration = ES.GetIntrinsic('%Temporal.Duration%');
-  const duration = new Duration(durationLike);
-  if (duration.years) throw new RangeError(`invalid duration field years`);
-  if (duration.months) throw new RangeError(`invalid duration field months`);
+    const result = Object.create(Absolute.prototype);
+    CreateSlots(result);
+    SetSlot(result, EPOCHNANOSECONDS, { ms, ns });
+    return result;
+  }
+  minus(durationLike = {}) {
+    const duration = ES.GetIntrinsic('%Temporal.duration%')(durationLike);
+    if (duration.years) throw new RangeError(`invalid duration field years`);
+    if (duration.months) throw new RangeError(`invalid duration field months`);
 
-  let delta = BigInt(duration.days) * 86400000000000n;
-  delta += BigInt(duration.hours) * 3600000000000n;
-  delta += BigInt(duration.minutes) * 60000000000n;
-  delta += BigInt(duration.seconds) * 1000000000n;
-  delta += BigInt(duration.milliseconds) * 1000000n;
-  delta += BigInt(duration.microseconds) * 1000n;
-  delta += BigInt(duration.nanosecond);
+    let { ms, ns } = GetSlot(this, EPOCHNANOSECONDS);
+    let negative = ms < 0 || ns < 0;
+    ns -= duration.nanoseconds;
```

Comment:
Should these property accesses be replaced by usages of GetSlot, given that we know this is a duration, but Duration.prototype may have been patched?

Reply:
YES!

Reply:
Changed to do that